ansible_runner/streaming.py

```diff
@@ -0,0 +1,220 @@
+        if _output is None:
+            _output = sys.stdout.buffer
+        self._input = _input
+        self._output = _output
+
+        self.kwargs = kwargs
+
+        private_data_dir = kwargs.get('private_data_dir')
+        if not private_data_dir:
+            private_data_dir = tempfile.TemporaryDirectory().name
+        self.private_data_dir = private_data_dir
+
+        self.status = "unstarted"
+        self.rc = None
+
+    def run(self):
+        while True:
+            line = self._input.readline()
+            data = json.loads(line)
+
+            if 'zipfile' in data:
+                zip_data = self._input.read(data['zipfile'])
+                buf = io.BytesIO(zip_data)
+                with zipfile.ZipFile(buf, 'r') as archive:
+                    archive.extractall(path=self.private_data_dir)
+            elif 'eof' in data:
+                break
+
+        kwargs_path = os.path.join(self.private_data_dir, 'kwargs')
+        if os.path.exists(kwargs_path):
+            with open(kwargs_path, "r") as kwf:
+                kwargs = json.load(kwf)
+            if not isinstance(kwargs, dict):
+                raise ValueError("Invalid kwargs data")
+        else:
+            kwargs = {}
+
+        self.kwargs.update(kwargs)
+
+        self.kwargs['quiet'] = True
+        self.kwargs['private_data_dir'] = self.private_data_dir
+        self.kwargs['status_handler'] = self.status_handler
+        self.kwargs['event_handler'] = self.event_handler
+        self.kwargs['artifacts_handler'] = self.artifacts_handler
+        self.kwargs['finished_callback'] = self.finished_callback
+
+        r = ansible_runner.interface.run(**self.kwargs)
+        self.status, self.rc = r.status, r.rc
+
+        # FIXME: do cleanup on the tempdir
+
+        return self.status, self.rc
+
+    def status_handler(self, status_data, runner_config):
+        self.status = status_data['status']
+        self._output.write(json.dumps(status_data).encode('utf-8'))
+        self._output.write(b'\n')
+        self._output.flush()
+
+    def event_handler(self, event_data):
+        self._output.write(json.dumps(event_data).encode('utf-8'))
+        self._output.write(b'\n')
+        self._output.flush()
+
+    def artifacts_handler(self, artifact_dir):
+        self._output.write(utils.stream_dir(artifact_dir))
+        self._output.flush()
+
+    def finished_callback(self, runner_obj):
+        self._output.write(json.dumps({'eof': True}).encode('utf-8'))
+        self._output.write(b'\n')
+        self._output.flush()
+        # self._output.close()
+
+
+class Processor(object):
+    def __init__(self, _input=None, status_handler=None, event_handler=None,
+                 artifacts_handler=None, cancel_callback=None, finished_callback=None, **kwargs):
+        if _input is None:
+            _input = sys.stdin.buffer
+        self._input = _input
+
+        self.kwargs = kwargs
+        self.config = ansible_runner.RunnerConfig(**kwargs)
```

Comment:
I don't see how kwargs would be enough to construct a RunnerConfig, at least not from the CLI options

```
$ ansible-runner process --help
usage: ansible-runner process [-h] [--version] [--debug] [--logfile LOGFILE]

optional arguments:
  -h, --help         show this help message and exit
  --version          show program's version number and exit
  --debug            enable ansible-runner debug output logging
                     (default=False)
  --logfile LOGFILE  log output messages to a file (default=None)
```

Additionally, all the parameters have already been specified in the original transmit(?) command.

Reply:
You're right, I'm not sure that this is meaningful or even not broken the way I have it here.  However, the handlers require a `RunnerConfig` object, so some kind of workaround is going to have to be found.

Reply:
I want to formalize my suggestions on this matter. I started out here not understanding what the requirements for the plugins are, and I have found these uses:

https://github.com/ansible/ansible-runner-http/blob/master/ansible_runner_http/events.py
https://github.com/ansible/ansible-runner-aio/blob/master/ansible_runner_aio/events.py
https://github.com/wtcross/ansible-runner-mqtt/blob/master/ansible_runner_mqtt/events.py

Those are all I can find on GitHub generally.

----
You had mentioned the idea of streaming back the details of the config object from the worker back to the controller. I think that's a bad idea. Looking further into it, I believe it wouldn't even _work_. In the http example, for example, it queries a setting like `runner_config.settings.get("runner_http_url", None)`, which uses `settings` as defined here.

https://github.com/ansible/ansible-runner/blob/b3b2be41340f1ae65899b8a504b5660cc8f1922b/ansible_runner/runner_config.py#L319

Observations from the example plugins:

a. The plugin never uses runner settings themselves, it uses plugin-specific settings
b. These settings can only be provided in the `env/settings` file

----
Wrapping this up, I don't _like_ the idea `ansible-runner process` accepting a positional argument for private_data_dir. But that's literally what the plugins require. There's no ambiguity. Maybe you could make it optional, because 99.99% of cases won't have any plugins and it's not needed in those places.

It's also established that plugins will be ran on the controller side. Combine the other precedents, that means that settings for the plugins need to be given in `env/settings` of the private_data_dir on the controller. Slap a bow on that and put it in docs.

This is all assuming that a CLI entrypoint for process is needed in the first place, and I don't feel like it should be needed for an initial merge.

Reply:
Also https://github.com/elmarx/ansible-runner-kafka

Reply:
"assuming that a CLI entrypoint for process is needed in the first place": this is a red herring.  The parameters are the same regardless of whether you invoke runner using the CLI or by importing it as a library.

Reply:
Sorry that wasn't very clear.

Where I was going with that is, we could have `ansible-runner transmit_and_process` as opposed to `transmit` and `process`, if we don't expect to use them independently of each other for the moment.

I think there is enough information here to implement the process functionality on its own, constructing the RunnerConfig locally. Even so, a combined transmit-and-process command would eliminate _any_ question about implementation because there is no need to re-construct a RunnerConfig for a previously-started job.